### PR TITLE
Make middleware selection more independent of build-time package availability

### PR DESCRIPTION
**Pull request by rotu**
Defer list of available RMWs to build time of downstream packages so you can build a dependent package even if RMW_IMPLEMENTATION is set to a value not available when the rmw_implementation package was built.

Remove CMake Config flag "RMW_IMPLEMENTATION_FORCE_DYNAMIC_LOADING".
Add CMake Config flag "RMW_IMPLEMENTATION_DISABLE_RUNTIME_SELECTION" so you can statically link a middleware, even if multiple present at build time. This value can be explicitly set, but by default preserves the behavior of "statically link if one middleware found".

Add a scary message if the user tries to change the implementation and built with RMW_IMPLEMENTATION_DISABLE_RUNTIME_SELECTION=ON

### CODE REVIEW

**Comment by rotu** [comment]
Alternative to #65

**Comment by rotu** [comment]
I’m not 100% sure, with the changes in this PR, if it’s still necessary to list middlewares in package.xml.

https://answers.ros.org/question/332359/why-does-rmw_implementation-have-a-build-dependency-on-certain-middleware-packages/

**Comment by dirk-thomas** [comment]
As mentioned on the other ticket (https://github.com/ros2/rmw_implementation/pull/65#issuecomment-528453114) imo it should be preserved that when only a single RMW implementation is present that by default `rmw_implementation` statically links against it. Since this patch changes that behavior I don't think that is a desired.

**Comment by rotu** [comment]
I believe that change is desirable. Behavior-changing optimizations such as direct vs indirect loading of RMW *should be* explicit rather than implicit.

The previous behavior is unintuitive (#58) and [undocumented](https://index.ros.org//doc/ros2/Tutorials/Working-with-multiple-RMW-implementations/).

If I’m tuning performance in real life, I’m going to want to compare multiple RMWs, and that means probably (1) having them checked out in my workspace at compile time or installed via apt, even if I’m only using one (2) toggling poco with the same RMW.

**Comment by dirk-thomas** [comment]
> (1) having them checked out in my workspace at compile time or installed via apt

For this the case of a single RMW impl at build time doesn't apply.

So in which scenario are you using a binary which was built with a single RMW implementation and then want to add other RMW implementations later? All binaries by Open Robotics use more than one RMW impl.

**Comment by rotu** [comment]
> since the user shouldn't "pay for" what they don't use / need

That’s the thing: you *can’t tell* when rmw_implementation is built whether the user will need dynamic linking when rmw_implementation is actually used in an overlay.

The previous behavior is an example of premature optimization.

In any case, if you install via apt, all users are forced to “pay for” this feature since the OSRF binaries are built with multiple typesupports.

> So in which scenario are you using a binary which was built with a single RMW implementation and then want to add other RMW implementations later?

A hypothetical: I’ve built my underlay to zoom under FastRTPS, and I’d like to try out CycloneDDS. Now I install FastRTPS alongside CycloneDDS, and inadvertently either change the FastRTPS performance by linking differently or accidentally compare statically linked FastRTPS to dynamically linked CycloneDDS.

**Comment by rotu** [comment]
Not to mention if you rosdep install, you get the other RMWs! So the previous behavior seems to discourage meaningful use of rosdep.

**Comment by dirk-thomas** [comment]
> That’s the thing: you can’t tell when rmw_implementation is built whether the user will need dynamic linking when rmw_implementation is actually used in an overlay.

Correct. That is why we have to make a decision at built time which might not be the right one in some cases. My point is that I think the current strategy aim to do the right thing for the common use cases.

> In any case, if you install via apt, all users are forced to “pay for” this feature since the OSRF binaries are built with multiple typesupports.

> Not to mention if you rosdep install, you get the other RMWs! So the previous behavior seems to discourage meaningful use of rosdep.

We simply decided that providing support for multiple RMW impl. in the provided binary packages is beneficial to the majority of the community.

> I’ve built my underlay to zoom under FastRTPS, and I’d like to try out CycloneDDS. Now I install FastRTPS alongside CycloneDDS, and inadvertently either change the FastRTPS performance by linking differently or accidentally compare statically linked FastRTPS to dynamically linked CycloneDDS.

I you want to measure performance (which is not a common use case compared to the overall community which just wants to use the binaries) you need to know what you are doing anyway. So if you don't want to compare apples with oranges you need to compile both RMW impl either each on their own or both together to have an exact comparison.

Let's just agree that we disagree on what each of us thinks the right choice is here. Maybe someone else wants to comment with their opinion to get a broader set of perspectives.

**Comment by rotu** [comment]
> Let's just agree that we disagree on what each of us thinks the right choice is here. Maybe someone else wants to comment with their opinion to get a broader set of perspectives.

:+1: @nuclearsandwich, could you weigh in?

**Comment by clalancette** [comment]
My opinion: we should default to always building dynamically, with an option to build one statically if the user wants.  That is, I think we should take this PR.  Most of the time, users are not concerned with absolute performance, and thus having dynamic loading be the default no matter how many DDS vendors you have installed seems to be the most consistent thing to do.

**Comment by nuclearsandwich** [comment]
> Not to mention if you rosdep install, you get the other RMWs! So the previous behavior seems to discourage meaningful use of rosdep.

I don't think that this has any bearing on the default linking behavior for rmw_implementations. Providing more complete group semantics and implementing them across the tooling is a separate (but valid) issue.

I agree with @dirk-thomas that anyone profiling will need to be paying attention to the rmw binding style and that which one is the default is less important. Our documentation for building from source focuses on the use of the default rmw. It might be worth adding something here: https://index.ros.org/doc/ros2/Installation/Dashing/Linux-Development-Setup/#install-more-dds-implementations-optional which mentions that if you first built a workspace with only one rmw implementation that you may need to reconfigure rmw_implementation to pick up the second one.

I also think defaulting to skipping poco makes sense as a default of "don't pay for what you aren't using" and yes someone may want to use it down the road and may not be fully aware of the change in behavior but I think that's a better situation than requiring people to opt-in to the optimization.

**Comment by rotu** [comment]
Even (especially) if the performance difference is extreme, we should provide a way to affirmatively opt in to static linking. There isn't one currently. ("Oops! I switched to another DDS but accidentally left the first RMW package installed with apt so it's building with dynamic linking and there's no way to tell").

Either it's important enough to make link type explicit or it's not. I don't see an argument for "this is important but not important enough to allow the user to turn ON static linking, only off".

**Comment by rotu** [comment]
Reverted to the static build if one RMW is found at build time. I think this is the wrong call, but at the end of the day, I don't think it will make much difference.

Please rereview

**Comment by rotu** [comment]
Rebased and revised commit message:

> 
> Defer list of available RMWs to build time of downstream packages so you can build a dependent package even if RMW_IMPLEMENTATION is set to a value not available when the rmw_implementation package was built.
> 
> Remove CMake Config flag "RMW_IMPLEMENTATION_FORCE_DYNAMIC_LOADING".
> Add CMake Config flag "RMW_IMPLEMENTATION_BAKED" so you can statically link a middleware, even if multiple present at build time. This value can be explicitly set, but by default preserves the behavior of "statically link if one middleware found".
> 
> Downgrade missing rmw library to a warning, since some RMW's like rmw_cyclonedds_cpp and rmw_fastrtps_dynamic_cpp (?) do not need to inject build dependencies.
> Add a scary message if the user tries to change the implementation and built with RMW_IMPLEMENTATION_BAKED=ON
> 
> Signed-off-by: Dan Rose <dan@digilabs.io>

File: rmw_implementation/CMakeLists.txt [hunk 1]
```diff
@@ -15,65 +15,50 @@ find_package(ament_cmake REQUIRED)
...
+if(count_available_rmw_implementations EQUAL 1)
+  set(rmw_implementation_baked_default ON)
+else()
+  set(rmw_implementation_baked_default OFF)
 endif()
 
-if(NOT RMW_IMPLEMENTATION_SUPPORTS_POCO)
-  message(STATUS "Dynamic loading disabled; directly referencing '${RMW_IMPLEMENTATION}'")
-  message(STATUS "")
+option(RMW_IMPLEMENTATION_BAKED "\
```
Comment (dirk-thomas) [inline rmw_implementation/CMakeLists.txt]: Instead of using "_BAKED" please use "_STATIC" based on the static vs. dynamic library loading.

File: rmw_implementation/CMakeLists.txt [hunk 2]
```diff
@@ -15,65 +15,50 @@ find_package(ament_cmake REQUIRED)
...
+  Use only the RMW implementation specified at build time, \
+  ignoring runtime value of the 'RMW_IMPLEMENTATION' environment variable \
+  and the 'RMW_IMPLEMENTATION' CMake option"
```
Comment (dirk-thomas) [inline rmw_implementation/CMakeLists.txt]: Doesn't `get_default_rmw_implementation()` take a CMake definition with that name into account when determining the default?

File: rmw_implementation/CMakeLists.txt [hunk 3]
```diff
@@ -15,65 +15,50 @@ find_package(ament_cmake REQUIRED)
...
+  message(STATUS "Dynamic loading disabled; directly referencing '${RMW_IMPLEMENTATION}'")
 else()
-  message(STATUS "Dynamic loading enabled with any of the available rmw implementations")
-  message(STATUS "")
+  # provides FindPoco.cmake and Poco on platforms without it
+  find_package(poco_vendor REQUIRED)
+  find_package(Poco REQUIRED COMPONENTS Foundation)
 
-  find_package(rmw REQUIRED)
+  message(STATUS "Dynamic loading enabled")
```
Comment (dirk-thomas) [inline rmw_implementation/CMakeLists.txt]: A more specific message would be better since the context might not be obvious for users, e.g.:

> Dynamic loading of RMW implementation enabled

Also above for the disabled case.

File: rmw_implementation/CMakeLists.txt [hunk 4]
```diff
@@ -15,65 +15,50 @@ find_package(ament_cmake REQUIRED)
...
 
-  link_directories(${Poco_LIBRARY_DIR})
```
Comment (dirk-thomas) [inline rmw_implementation/CMakeLists.txt]: Why was this removed?

File: rmw_implementation/CMakeLists.txt [hunk 5]
```diff
@@ -15,65 +15,50 @@ find_package(ament_cmake REQUIRED)
...
   add_library(${PROJECT_NAME} SHARED
     src/functions.cpp)
   ament_target_dependencies(${PROJECT_NAME}
     "Poco"
     "rmw")
   target_compile_definitions(${PROJECT_NAME}
     PUBLIC "DEFAULT_RMW_IMPLEMENTATION=${RMW_IMPLEMENTATION}")
-  target_link_libraries(${PROJECT_NAME})
```
Comment (dirk-thomas) [inline rmw_implementation/CMakeLists.txt]: This change seems to be unrelated. It would be easier if it would be done in a separate PR (not necessary now but in the future).

File: rmw_implementation/rmw_implementation-extras.cmake.in [hunk 6]
```diff
@@ -15,69 +15,57 @@
...
+  set(requested_rmw_implementation "${RMW_IMPLEMENTATION}")
+  set(requested_rmw_implementation_from "CMake")
+elseif(NOT "$ENV{RMW_IMPLEMENTATION}" STREQUAL "")
+  set(requested_rmw_implementation "$ENV{RMW_IMPLEMENTATION}")
+  set(requested_rmw_implementation_from
+    "environment variable 'RMW_IMPLEMENTATION'")
+else()
+  set(requested_rmw_implementation "@RMW_IMPLEMENTATION@")
+  set(requested_rmw_implementation_from
+    "the default when @PROJECT_NAME@ wasbuilt")
```
Comment (dirk-thomas) [inline rmw_implementation/rmw_implementation-extras.cmake.in]: Nitpick: space between "was" and "built".

File: rmw_implementation/rmw_implementation-extras.cmake.in [hunk 7]
```diff
@@ -15,69 +15,57 @@
...
-        "The RMW implementation has been specified as "
-        "'${requested_rmw_implementation}' "
-        "through the environment variable 'RMW_IMPLEMENTATION', "
-        "however this needs to match the RMW implementation "
-        "'${default_rmw_implementation}', "
-        "which was specified when the 'rmw_implementation' package was built.")
-    endif()
+if(@RMW_IMPLEMENTATION_BAKED@)
+  message(STATUS "Using RMW implementation '@RMW_IMPLEMENTATION@'")
+  if(NOT requested_rmw_implementation STREQUAL @RMW_IMPLEMENTATION@)
```
Comment (dirk-thomas) [inline rmw_implementation/rmw_implementation-extras.cmake.in]: Please add quotes around `@RMW_IMPLEMENTATION@`.

File: rmw_implementation/rmw_implementation-extras.cmake.in [hunk 8]
```diff
@@ -15,69 +15,57 @@
...
+    message(ERROR
+      "The RMW implementation has been specified as "
+      "'${requested_rmw_implementation}' via "
+      "${requested_rmw_implementation_from}, but rmw_implementation was built "
+      "without dynamic linking enabled. We will use '@RMW_IMPLEMENTATION@'."
```
Comment (dirk-thomas) [inline rmw_implementation/rmw_implementation-extras.cmake.in]: Nitpick: dynamic loading - not linking.

File: rmw_implementation/rmw_implementation-extras.cmake.in [hunk 9]
```diff
@@ -15,69 +15,57 @@
...
-        "'${requested_rmw_implementation}' "
-        "through the environment variable 'RMW_IMPLEMENTATION', "
-        "however this needs to match the RMW implementation "
-        "'${default_rmw_implementation}', "
-        "which was specified when the 'rmw_implementation' package was built.")
-    endif()
+if(@RMW_IMPLEMENTATION_BAKED@)
+  message(STATUS "Using RMW implementation '@RMW_IMPLEMENTATION@'")
+  if(NOT requested_rmw_implementation STREQUAL @RMW_IMPLEMENTATION@)
+    message(ERROR
```
Comment (dirk-thomas) [inline rmw_implementation/rmw_implementation-extras.cmake.in]: Does it make sense to continue in this case? We can't satisfy the explicitly requested configuration. I would argue a `FATAL_ERROR` message would be more appropriate here. Btw. that was also the behavior before.

File: rmw_implementation/rmw_implementation-extras.cmake.in [hunk 10]
```diff
@@ -15,69 +15,57 @@
...
+  list(APPEND rmw_implementation_INCLUDE_DIRS
+    ${@RMW_IMPLEMENTATION@_INCLUDE_DIRS})
+  list(APPEND rmw_implementation_LIBRARIES
+    ${@RMW_IMPLEMENTATION@_LIBRARIES})
+else()
+
+  get_available_rmw_implementations(available_rmw_implementations)
+
+  if(NOT requested_rmw_implementation IN_LIST available_rmw_implementations)
+    message(WARNING
```
Comment (dirk-thomas) [inline rmw_implementation/rmw_implementation-extras.cmake.in]: Same as above: this was a fatal error and should stay one.

File: rmw_implementation/rmw_implementation-extras.cmake.in [hunk 11]
```diff
@@ -15,69 +15,57 @@
...
-  list(APPEND rmw_implementation_DEFINITIONS ${${selected_rmw_implementation}_DEFINITIONS})
-  list(APPEND rmw_implementation_INCLUDE_DIRS ${${selected_rmw_implementation}_INCLUDE_DIRS})
-  list(APPEND rmw_implementation_LIBRARIES ${${selected_rmw_implementation}_LIBRARIES})
-endif()
 
-find_package(Threads REQUIRED)
-list(APPEND rmw_implementation_LIBRARIES "${CMAKE_THREAD_LIBS_INIT}")
+  find_package(Threads REQUIRED)
+  list(APPEND rmw_implementation_LIBRARIES "${CMAKE_THREAD_LIBS_INIT}")
+endif()
```
Comment (dirk-thomas) [inline rmw_implementation/rmw_implementation-extras.cmake.in]: Readd newline at the EOF.

File: rmw_implementation/CMakeLists.txt (hunk 1 above)
Comment (rotu) [reply rmw_implementation/CMakeLists.txt]: There is a reason I used "Baked" instead. "static" means many things (static class, static variable, static linkage), and the important part is the RMW Implementation is fixed at the time of build and unchanging; i.e. "baked in" at time of build. If you have a better idea than "static" I could get on board.

Comment (dirk-thomas) [reply rmw_implementation/CMakeLists.txt]: `STATICALLY_LINKED` sounds fine to me if you think "static" is too generic.

File: rmw_implementation/CMakeLists.txt (hunk 2 above)
Comment (rotu) [reply rmw_implementation/CMakeLists.txt]: Yes. By means of the build-time value of RMW_IMPLEMENTATION environment variable and CMake option. How should I better phrase this?

https://github.com/ros2/rmw/blob/63b71a1f7188b78bea6a893bbb0b6845c3dd3a80/rmw_implementation_cmake/cmake/get_default_rmw_implementation.cmake#L32-L54

File: rmw_implementation/CMakeLists.txt (hunk 4 above)
Comment (rotu) [reply rmw_implementation/CMakeLists.txt]: Because it's redundant with ament_target_dependencies(... "Poco" ...)

File: rmw_implementation/CMakeLists.txt (hunk 5 above)
Comment (rotu) [reply rmw_implementation/CMakeLists.txt]: If it's not worth writing up an issue over, I generally don't regard it as important enough for its own PR. I'll consider it in the future

File: rmw_implementation/CMakeLists.txt (hunk 2 above)
Comment (dirk-thomas) [reply rmw_implementation/CMakeLists.txt]: I am not sure how to phrase it better but currently `and the 'RMW_IMPLEMENTATION' CMake option` suggests that you can't use `RMW_IMPLEMENTATION` to influence this which isn't true for this package.

File: rmw_implementation/CMakeLists.txt (hunk 4 above)
Comment (dirk-thomas) [reply rmw_implementation/CMakeLists.txt]: Why would that be redundant? `ament_target_dependencies` doesn't consider `*_LIBRARY_DIR` variables and if `*_LIBRARIES` only contains a name (not an absolute path) this will fail without the link directory.

Comment (rotu) [reply rmw_implementation/CMakeLists.txt]: Wait - so then where does it get added to the link_libraries or target_link_libraries? I would expect Poco_LIBRARY_DIR to be set there?

File: rmw_implementation/CMakeLists.txt (hunk 1 above)
Comment (dirk-thomas) [reply rmw_implementation/CMakeLists.txt]: Just for the record: you chose the suffix `_DISABLE_RUNTIME_SELECTION`.

File: rmw_implementation/CMakeLists.txt (hunk 4 above)
Comment (dirk-thomas) [reply rmw_implementation/CMakeLists.txt]: `link_directories()` like `include_directories()` changes global state in CMake and is being considered by every target.

File: rmw_implementation/rmw_implementation-extras.cmake.in [hunk 12]
```diff
@@ -15,69 +15,57 @@
...
+
+  get_available_rmw_implementations(available_rmw_implementations)
+
+  if(NOT requested_rmw_implementation IN_LIST available_rmw_implementations)
+    message(FATAL_ERROR
+      "The RMW implementation has been specified as "
+      "'${requested_rmw_implementation}' via "
+      "${requested_rmw_implementation_from}, but it is not available at this "
+      "time.\n Make sure it does not expect a typesupport or other build "
+      "dependencies.\n\n Currently available middlewares: "
```
Comment (dirk-thomas) [inline rmw_implementation/rmw_implementation-extras.cmake.in]: Nitpick: no space after the newline.

Same above.

File: rmw_implementation/rmw_implementation-extras.cmake.in [hunk 13]
```diff
@@ -15,69 +15,57 @@
...
+else()
+
+  get_available_rmw_implementations(available_rmw_implementations)
+
+  if(NOT requested_rmw_implementation IN_LIST available_rmw_implementations)
+    message(FATAL_ERROR
+      "The RMW implementation has been specified as "
+      "'${requested_rmw_implementation}' via "
+      "${requested_rmw_implementation_from}, but it is not available at this "
+      "time.\n Make sure it does not expect a typesupport or other build "
```
Comment (dirk-thomas) [inline rmw_implementation/rmw_implementation-extras.cmake.in]: > Make sure it does not expect a typesupport or other build dependencies.

I don't understand the meaning / purpose of this part of the message. The specified RMW impl isn't available - what does the type support and/or other dependencies have to do with that?

File: rmw_implementation/CMakeLists.txt (hunk 4 above)
Comment (rotu) [reply rmw_implementation/CMakeLists.txt]: Okay so where does Poco get added to link_libraries? It should be added to link_directories as close to that point as possible.

Comment (dirk-thomas) [reply rmw_implementation/CMakeLists.txt]: By the `ament_target_dependencies` call.

Comment (rotu) [reply rmw_implementation/CMakeLists.txt]: Which for some reason *doesn’t* also add it to LINK_DIRECTORIES?

Comment (dirk-thomas) [reply rmw_implementation/CMakeLists.txt]: Using `link_directories` is not recommended imo because it exposes you to the same problem as global include directories: adding one might affect the lookup of any other library. Therefore the ament function doesn't use it. Also only very few packages use that (imo outdated) approach.

File: rmw_implementation/rmw_implementation-extras.cmake.in (hunk 13 above)
Comment (rotu) [reply rmw_implementation/rmw_implementation-extras.cmake.in]: I wanted to make sure that things like rmw_fastrtps_dynamic_cpp and rmw_cyclonedds_cpp were not rejected by the build system if they were unavailable at the time this was built. I think you’re right that, now that this is a fatal error, that language is unnecessary

File: rmw_implementation/CMakeLists.txt (hunk 4 above)
Comment (rotu) [reply rmw_implementation/CMakeLists.txt]: Okay, so if it's the wrong approach, why are you asking me to use it? I'm so confused :-/. If it's needed, it seems ament_target_dependencies should set it up, as it is part of the dependency. If it's not needed, then neither CMakeLists nor ament_target_dependencies should be setting it.

Also, ${Poco_LIBRARY_DIR} doesn't even seem to be defined when CMake gets to that line, so thhis all may be a moot point.

Comment (dirk-thomas) [reply rmw_implementation/CMakeLists.txt]: `ament_target_dependencies` doesn't use it since it generates an order problem. That is why this package sets it explicitly using `link_directories` even though this is not great. This package takes the responsibility that it doesn't break the order of searched libraries.

As an alternative you could use the library dir to search for the library name explicitly using `find_library()` and then use the absolute path of the found library.

> ${Poco_LIBRARY_DIR} doesn't even seem to be defined when CMake gets to that line

You would need to verify that on all targeted platforms / supported Poco versions. Also a reference when this has been removed from Poco would be beneficial.

Anyway neither of this is related to the change you are proposing in this PR. Therefore I recommend to not change the way this is currently being done. And if you like to modify it do so in a separate PR where it can be iterated on independently.

File: rmw_implementation/CMakeLists.txt [hunk 14]
```diff
@@ -15,67 +15,55 @@ find_package(ament_cmake REQUIRED)
...
 
+  message(STATUS "Runtime selection of RMW enabled")
+  
   link_directories(${Poco_LIBRARY_DIR})
   add_library(${PROJECT_NAME} SHARED
     src/functions.cpp)
   ament_target_dependencies(${PROJECT_NAME}
     "Poco"
     "rmw")
+
```
Comment (dirk-thomas) [inline rmw_implementation/CMakeLists.txt]: Please revert unnecessary changes like this.

File: rmw_implementation/CMakeLists.txt [hunk 15]
```diff
@@ -15,67 +15,55 @@ find_package(ament_cmake REQUIRED)
...
   target_compile_definitions(${PROJECT_NAME}
     PUBLIC "DEFAULT_RMW_IMPLEMENTATION=${RMW_IMPLEMENTATION}")
-  target_link_libraries(${PROJECT_NAME})
   configure_rmw_library(${PROJECT_NAME})
-
```
Comment (dirk-thomas) [inline rmw_implementation/CMakeLists.txt]: Please revert unnecessary changes like this.

File: rmw_implementation/CMakeLists.txt [hunk 16]
```diff
@@ -15,67 +15,55 @@ find_package(ament_cmake REQUIRED)
...
-find_package(poco_vendor)
-find_package(Poco COMPONENTS Foundation)
 find_package(rmw_implementation_cmake REQUIRED)
 
 if(BUILD_TESTING)
   find_package(ament_lint_auto REQUIRED)
   ament_lint_auto_find_test_dependencies()
 endif()
 
+find_package(rmw REQUIRED)
```
Comment (dirk-thomas) [inline rmw_implementation/CMakeLists.txt]: Why is this added here? The following function `get_default_rmw_implementation` is provided by `rmw_implementation_cmake` which was found earlier. If it is needed further down maybe moving it closer to where it is used?

File: rmw_implementation/CMakeLists.txt (hunk 4 above)
Comment (rotu) [reply rmw_implementation/CMakeLists.txt]: done. #70

File: rmw_implementation/rmw_implementation-extras.cmake.in [hunk 17]
```diff
@@ -15,69 +15,55 @@
...
-  set(selected_rmw_implementation "${default_rmw_implementation}")
-  message(STATUS "Using RMW implementation '${selected_rmw_implementation}' as default")
-  find_package("${selected_rmw_implementation}" REQUIRED)
 
-  # TODO should never need definitions and include dirs?
-  list(APPEND rmw_implementation_DEFINITIONS ${${selected_rmw_implementation}_DEFINITIONS})
-  list(APPEND rmw_implementation_INCLUDE_DIRS ${${selected_rmw_implementation}_INCLUDE_DIRS})
-  list(APPEND rmw_implementation_LIBRARIES ${${selected_rmw_implementation}_LIBRARIES})
+  find_package(Threads REQUIRED)
+  list(APPEND rmw_implementation_LIBRARIES "${CMAKE_THREAD_LIBS_INIT}")
```
Comment (dirk-thomas) [inline rmw_implementation/rmw_implementation-extras.cmake.in]: Why is this local to the `else` case where it was also used in the `if` case before?

File: rmw_implementation/rmw_implementation-extras.cmake.in [hunk 18]
```diff
@@ -15,69 +15,55 @@
...
+
+  # TODO should never need definitions and include dirs?
+  list(APPEND rmw_implementation_DEFINITIONS
+    ${@RMW_IMPLEMENTATION@_DEFINITIONS})
+  list(APPEND rmw_implementation_INCLUDE_DIRS
+    ${@RMW_IMPLEMENTATION@_INCLUDE_DIRS})
+  list(APPEND rmw_implementation_LIBRARIES
+    ${@RMW_IMPLEMENTATION@_LIBRARIES})
+else()
+
```
Comment (dirk-thomas) [inline rmw_implementation/rmw_implementation-extras.cmake.in]: Nitpick: please remove the empty line in the beginning of this block.

File: rmw_implementation/CMakeLists.txt (hunk 16 above)
Comment (rotu) [reply rmw_implementation/CMakeLists.txt]: Ah you're right.

Comment (rotu) [reply rmw_implementation/CMakeLists.txt]: Wait... where does this even need the dependency on rmw? It seems like we don't even use it

Comment (rotu) [reply rmw_implementation/CMakeLists.txt]: nevermind. I see

File: rmw_implementation/rmw_implementation-extras.cmake.in (hunk 17 above)
Comment (rotu) [reply rmw_implementation/rmw_implementation-extras.cmake.in]: You're totally right. I'm not sure that the threading library even belongs in here but removing it breaks the build, so :man_shrugging:

**Comment by rotu** [comment]
Rebased, reworded, and cleaned up unnecessary whitespace changes, as well as functional issues you pointed out. Thanks for your patience, @dirk-thomas!

File: rmw_implementation/CMakeLists.txt (hunk 5 above)
Comment (dirk-thomas) [reply rmw_implementation/CMakeLists.txt]: > I generally don't regard it as important enough for its own PR.

The problem simply is discoverability: if you see a commit / PR labeled "foo" you don't expect it to change "bar". That makes tracing changes in the future more difficult. Also making a separate PR (not needed in this case) literally takes less than a minute - no need to create an issue for it.

File: rmw_implementation/CMakeLists.txt (hunk 15 above)
Comment (dirk-thomas) [reply rmw_implementation/CMakeLists.txt]: Still pending.

File: rmw_implementation/CMakeLists.txt [hunk 19]
```diff
@@ -91,8 +78,6 @@ register_rmw_implementation(
   "cpp:rosidl_typesupport_cpp"
 )
 
-# necessary since get_available_rmw_implementations excludes rmw_implementation
-list(APPEND RMW_IMPLEMENTATIONS "rmw_implementation")
```
Comment (dirk-thomas) [inline rmw_implementation/CMakeLists.txt]: Can you describe why this was removed?

File: rmw_implementation/CMakeLists.txt (hunk 15 above)
Comment (rotu) [reply rmw_implementation/CMakeLists.txt]: Oops. I gotta get better at gitting :-/

Comment (dirk-thomas) [reply rmw_implementation/CMakeLists.txt]: Please just add a commit to the PR in the future. Now I either have to re-review the whole patch or just assume what was changed during the forced push.

File: rmw_implementation/CMakeLists.txt (hunk 19 above)
Comment (rotu) [reply rmw_implementation/CMakeLists.txt]: It's not needed since `rmw-implementation-extras.cmake.in` no longer references `@RMW_IMPLEMENTATIONS@`

**Comment by dirk-thomas** [comment]
CI builds with only FastRTPS:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/8227/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/4199/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/6708/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/8112/)

CI builds with FastRTPS and Connext:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/8228/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/4200/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/6709/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/8113/)

**Comment by dirk-thomas** [comment]
Thanks for this improvement and for iterating on the patch!
